README
Tabs2Text
=========

*Tabs2Text* is a Firefox extension (add-on) to export all your opened tabs to a text file (CSV, JSON, Markdown). It exports title and URL's page.


CHANGELOG
2020.06.11:
    * Fix major bugs
    * Remove unused code (cleaning code)
    * Add localization for:
        - en, en_GB, en_US, pt_BR, pt_PT
    * Include HTML as an export option
    * Add preferences panel to remember user options (options are saved locally)
    * Release for Google Chrome

2020.04.24:
    * Release for Mozilla Firefox
    * First version


tabs2text.js
/************************
 ** TABS2TEXT
 **   description: Export all your opened tabs to a CSV, JSON or Markdown file.
 **   homepage_url: https://github.com/deniszanin/tabs2text
 **
 ** FILE: tabs2text.js
 ************************
 *
 * =====================
 * = GLOBALS VARIABLES =
 **/
const DEFAULT_FILE_PREFIX = 'tabs2text';
const DEFAULT_GITHUB_PAGE = 'https://github.com/deniszanin/tabs2text';
const DEFAULT_CURRENT_WINDOW_TABS_ONLY = false;
const DEFAULT_FILE_TYPE = "CSV";

let g_fileExtension = "txt";
let g_fileFormat;
let g_isCurrentWindow;
let g_objJSONContent = [];

/**
 * =======================
 * = PROTOTYPE functions =
 **
 **
 * Function to format Date object to a specific string.
 * @source: https://stackoverflow.com/questions/3066586/get-string-in-yyyymmdd-format-from-js-date-object
 * @param {*}
 * @returns {string} strDate
 **/
Date.prototype.yyyymmdd = function() {
  var mm = this.getMonth() + 1; // getMonth() is zero-based
  var dd = this.getDate();

  return [this.getFullYear(),
          (mm>9 ? '' : '0') + mm,
          (dd>9 ? '' : '0') + dd
         ].join('');
};

/**
 * =============
 * = FUNCTIONS =
 **
 * Function to load addon settings and add event listeners.
 * @param {*}
 * @return {*}
 **/
function loadSettings() {
  // Add an event listener 'click' to GENERATE button.
  let h_btnGenerate = document.getElementById('tabs-generateCSV');
  h_btnGenerate.addEventListener('click', function() {
    exportFile(_getUIFormat(), _getUIALLTabs());
  }, false);

  // Add an event listener 'click' to ABOUT button.
  let h_btnAbout = document.getElementById('tabs-aboutEXT');
  h_btnAbout.addEventListener('click', function() {
    let tabAbout = browser.tabs.create({
      url:DEFAULT_GITHUB_PAGE
    });
    tabAbout.then();
  }, false);

  // Add an event listener 'onChange' to EXPORT ALLTABS RADIO.
  let h_rdioExport = document.getElementById('export_allTabs_yes');
  h_rdioExport.addEventListener("RadioStateChange", function(radioEvent) {
    let thisRadio = event.target;

    if (thisRadio.checked)
      g_isCurrentWindow = true;
    else
      g_isCurrentWindow = false;
  }, false);

  //////
    /////// 
      ////////
  ////// LOADING SETTINGS

  // User preferences to global variables.
  g_fileFormat = _getUIFormat() || DEFAULT_FILE_TYPE;
  g_isCurrentWindow = _getUIALLTabs() || DEFAULT_CURRENT_WINDOW_TABS_ONLY;

  // Update user interface version.
  _setUIVersion();
  _setUIIcon();

  if (g_isCurrentWindow)
    document.getElementById('export_allTabs_yes').checked = true;
  else
    document.getElementById('export_allTabs_yes').checked = false;
}

/**
 * Function to export tabs information to a file.
 * @param {string} argFileFormat
 * @param {boolean} argCurrentWindow
 * @returns {*}
 **/
function exportFile(argFileFormat, argCurrentWindow) {
  if (argFileFormat == "") throw new Error ("Error: invalid exportFile parameter.");

  _setFileExtension(argFileFormat);
  _getBrowserTabs(argCurrentWindow).then((objALLTabs) => {
    let idWindow = "ALL";
    let tmpContent = "";
    let tmpCounter = 1;

    for (let currentTab of objALLTabs) {
      tmpContent = tmpContent + formatContent(argFileFormat, tmpCounter, currentTab.title, currentTab.url, formatDate(currentTab.lastAccessed));
      tmpCounter += 1;

      if (argCurrentWindow) idWindow = currentTab.windowId;
    }

    let fileContent = formatHeader(argFileFormat);
    fileContent = fileContent + tmpContent;

    if (argFileFormat == "JSON") {
      fileContent = JSON.stringify(g_objJSONContent, null, 4);
    }

    let fileBlob = new Blob([fileContent], { type: 'text/plain;charset=utf=8' });
    let tmpURL = (window.webkitURL || window.URL).createObjectURL(fileBlob);

    let nowDate = _getDate();
    let expFilename = DEFAULT_FILE_PREFIX + "-" + nowDate + "-" + idWindow + "." + g_fileExtension;
    ext_NotifyDownloader(tmpURL, expFilename, fileBlob);
  });
}

/**
 * Function to format a given date.
 * @source: https://stackoverflow.com/questions/23593052/format-javascript-date-as-yyyy-mm-dd
 * @param {string} argDate
 * @returns {string} retDate
 **/
function formatDate(argDate) {
  let tDate = new Date(argDate),
    month = '' + (tDate.getMonth() + 1),
    day = '' + tDate.getDate(),
    year = tDate.getFullYear(),
    hours = '' + tDate.getHours(),
    minutes = '' + tDate.getMinutes();

    if (month.length < 2) month = '0' + month;
    if (day.length < 2) day = '0' + day;
    if (hours.length < 2) hours = '0' + hours;
    if (minutes.length < 2) minutes = '0' + minutes;

    return [month, day, year, hours, minutes].join('-');
}

/**
 * Function to format the header of the file.
 * @param {string} argFormat
 * @returns {string} returnContent
 **/
function formatHeader(argFormat) {
  let returnContent = "";

  if (argFormat == 'CSV') {
    returnContent = "id;title;url;last acessed\n";
  } else if (argFormat == 'MD') {
    returnContent = "TABS2TEXT LINKS\n";
    returnContent = returnContent + "================\n";
    returnContent = returnContent + "# LINKS\n\n";
    returnContent = returnContent + "## Generated by [Tabs2Text](" + DEFAULT_GITHUB_PAGE + ")\n\n";
  } else if (argFormat == 'JSON') {
    returnContent = "";
  }

  return returnContent;
}

/**
 * Function to format the content of tabs, which will be exported to the file.
 * @param {string} argFormat
 * @param {int} argID
 * @param {string} argTitle
 * @param {string} argURL
 * @param {string} argLastAccessed
 **/
function formatContent(argFormat, argID, argTitle, argURL, argLastAccessed) {
  let returnContent = "";
  let t_objContent = {};

  if (argFormat == 'CSV') {
    returnContent = argID + ";" + argTitle + ";" + argURL + ";" + argLastAccessed + "\n";
  } else if (argFormat == 'MD') {
    returnContent = "### " + argID + ". [" + argTitle + "](" + argURL + "), last accessed: " + argLastAccessed + "\n";
  } else if (argFormat == 'JSON') {
    t_objContent.id = argID;
    t_objContent.title = argTitle;
    t_objContent.url = argURL;
    t_objContent.last = argLastAccessed;

    g_objJSONContent.push(t_objContent);
  }

  return returnContent;
}

/**
 * =========================
 * = GET and SET functions =
 **
 **
 * Function to retrieve information about browser tabs.
 * @param {boolean} argCurrentWindow
 * @returns {object} retTabs
 **/
function _getBrowserTabs(argCurrentWindow) {
  if (argCurrentWindow) return browser.tabs.query({currentWindow: true});
  if (!argCurrentWindow) return browser.tabs.query({});
}

/**
 * Function to get current date and format it.
 * @param {*}
 * @returns {*}
 **/
function _getDate() {
  let tDate = new Date();
  return tDate.yyyymmdd();
}

/**
 * Function to get User Interface (UI) "file format" option.
 * @param {*}
 * @returns {string} retFormat
 **/
function _getUIFormat() {
  let objOptions = document.getElementsByName('format_file_radio');
  let intOptions = objOptions.length;
  let retFormat = "";

  for (let c=0;c < intOptions; c++) {
    if (objOptions[c].checked == true) retFormat = objOptions[c].value;
  }

  return retFormat;
}

/** 
 * Function to get User Interface (UI) "ALL tabs" option (checked or not).
 * @param {*}
 * @returns {boolean} retCurrentWindow
 **/
function _getUIALLTabs() {
  let objALLTabs = document.getElementsByName('export_allTabs_radio');
  let intALLTabs = objALLTabs.length;
  let retCurrentWindow = DEFAULT_CURRENT_WINDOW_TABS_ONLY;

  for (let c=0;c < intALLTabs; c++) {
    if (objALLTabs[c].checked == true) {
      retCurrentWindow = (objALLTabs[c].value === "true");
    } 
  }

  return retCurrentWindow;
}

/**
 * Function to set the current version of this extension (addon) in UI.
 * @param {*}
 * @returns {*}
 **/
function _setUIVersion() {
  let locManifest = browser.runtime.getManifest();
  let spnVersion = document.getElementById("v-current-version");
  spnVersion.innerText = locManifest.version;
}

/**
 * Function to set the icon of this extension (addon) in UI.
 * @param {*}
 * @returns {*}
 **/
function _setUIIcon() {
  let locManifest = browser.runtime.getManifest();
  let srcIcon = document.getElementById("ico-tabs2text");
  srcIcon.src = locManifest.icons["48"];
}

/**
 * Function to set file extension.
 * @param {string} argFormat
 * @returns {*}
 **/
function _setFileExtension(argFormat) {
  if (argFormat == 'CSV') {
    g_fileExtension = "csv";
  } else if (argFormat == 'MD') {
    g_fileExtension = "md";
  } else if (argFormat == 'JSON') {
    g_fileExtension = "json";
  } else {
    g_fileExtension = "txt";
  }
}

/**
 * ===================
 * = EXTERNAL NOTIFY =
 **
 **
 * Function to notify and send commands to 'background.js'.
 * @param {string} argURL
 * @param {string} argFilename
 * @param {blog} argBlob
 * @returns {*}
 */
function ext_NotifyDownloader(argURL, argFilename, argBlob) {
  browser.runtime.sendMessage({"url": argURL, "filename": argFilename, "blob": argBlob});
}

/**
 * ===================
 * = EVENT LISTENERS =
 **/
 // Default action on loading.
 document.addEventListener("DOMContentLoaded", loadSettings);

 //EOF


background.js
/************************
 ** TABS2TEXT
 **   description: Export all your opened tabs to a CSV, HTML, JSON or Markdown file.
 **   homepage_url: https://github.com/deniszanin/tabs2text
 **   license: Creative Commons Attribution-NonCommercial 4.0 International (CC BY-NC 4.0)
 **
 ** FILE: background.js
 ************************
 *
 **/
"use strict";

/**
 * =====================
 * = GLOBALS VARIABLES =
 **/
const DEFAULT_STORAGE_OBJECT = chrome.storage.local;
const DEFAULT_CURRENT_WINDOW_TABS_ONLY = false;
const DEFAULT_FILE_TYPE = "HTML";

// For debugging application
const DEBUG_MODE = false; // Change it to "true", for debug mode.
if (DEBUG_MODE) console.log("DEBUG: TABS2TEXT DEBUG is ON.");

// Refreshing the tabs badge counter
updateBadgeCount();

/**
 * =============
 * = FUNCTIONS =
 **
 **
 * Function to refresh the tabs badge counter.
 * @source: https://github.com/mdn/webextensions-examples/blob/master/tabs-tabs-tabs/tabs.js
 * @param {} tabId
 * @returns {*}
 **/
function updateBadgeCount(tabId)
{
	chrome.tabs.query({}, function(objTabs) {
		let length = objTabs.length;

		chrome.browserAction.setBadgeText( {text: length.toString()} );
		chrome.browserAction.setBadgeBackgroundColor({"color": "blue"});
	});
}

/**
 * Function to receive an file content and save it to the local file system.
 * @param {string} argURL
 * @param {string} argFilename
 * @param {blob} argBlob
 * @returns {*}
 **/
function downloadFileAndSaveAs(argURL, argFilename, argBlob)
{
	let sanURL = argURL;

	// Firefox bug fix: create an URL again, based on argBlob.
	if (typeof(browser) != "undefined") sanURL = (window.webkitURL || window.URL).createObjectURL(argBlob);

	let downloadOptions = {
		url: sanURL,
		filename : argFilename,
		conflictAction : "uniquify",
		saveAs : true
	};

	chrome.downloads.download(downloadOptions, downloadID => {
		if (DEBUG_MODE) console.log(`DEBUG: Started downloading, id ${downloadID}, ok.`);
	});
}

 /**
  * Function to get external notifications (signals) and messages from tabs2text.js file.
  * @param {object} argEXTMessages
  * @returns {*}
  **/
function receiveFileContentMessage(argMessage) {
	if (DEBUG_MODE) console.log(`DEBUG: Message/notification received, message ${JSON.stringify(argMessage)}.`);

	downloadFileAndSaveAs(argMessage.url, argMessage.filename, argMessage.blob);
}

/**
 * ===================
 * = EVENT LISTENERS =
 *
 ** Default action on tabs creation or destruction.
 ** Default action when receiveing messages from pages.
 ** Default action on extension install.
 **/
chrome.tabs.onRemoved.addListener( (tabId) => { updateBadgeCount(tabId); });
chrome.tabs.onCreated.addListener( (tabId) => { updateBadgeCount(tabId); });
chrome.runtime.onMessage.addListener(function(message, sender, sendResponse) {
	const responses = {
		debug_log_on: () => {
			if (DEBUG_MODE) console.log("DEBUG-EXT: " + message.object);
		},
		debug_warn_on: () => {
			if (DEBUG_MODE) console.warn("DEBUG-EXT: " + message.object);
		},
		to_downloader: () => {
			receiveFileContentMessage(message.object);
		}
	};

	if (message.type in responses) {
		return responses[message.type]();
	}
});

// Fill BrowserStorage with default values.
chrome.runtime.onInstalled.addListener(function() {
	const defaultSettings = {
		storCurrentWindow: DEFAULT_CURRENT_WINDOW_TABS_ONLY,
		storFileFormat: DEFAULT_FILE_TYPE
	};

	// Store initial values in BrowserStorage.
	DEFAULT_STORAGE_OBJECT.set(defaultSettings, () => {
		if (DEBUG_MODE) console.log(`DEBUG: Initial values set in BrowserStorage.`);
	});
});

// #TODO
chrome.commands.onCommand.addListener(function(command)
{
	if (command == "t2t-export") {
		if (DEBUG_MODE) console.log("DEBUG: #TODO: Shortcut, exporting file.");
	}
});

// EOF

tabs2text-utils.js
/************************
 ** TABS2TEXT
 **   description: Export all your opened tabs to a CSV, HTML, JSON or Markdown file.
 **   homepage_url: https://github.com/deniszanin/tabs2text
 **   license: Creative Commons Attribution-NonCommercial 4.0 International (CC BY-NC 4.0)
 **
 ** FILE: tabs2text-utils.js
 ************************
 **/
"use strict";

/**
 * ====================
 * = EVENT LISTENERS  =
 * = LOCALES function =
 **
 ** Default action on loading page.
 **/
document.addEventListener("DOMContentLoaded", () => {
	for (const element of document.querySelectorAll("[data-i18n")) {
		element.innerText = chrome.i18n.getMessage(element.getAttribute("data-i18n")).replace(/&quot;/g,"\"");
	}
});

/**
 * =======================
 * = PROTOTYPE functions =
 **
 **
 * Function to format Date object to a specific string.
 * @source: https://stackoverflow.com/questions/3066586/get-string-in-yyyymmdd-format-from-js-date-object
 * @param {*}
 * @returns {string} strDate
 **/
Date.prototype.yyyymmdd = function()
{
	var mm = this.getMonth() + 1; // getMonth() is zero-based
	var dd = this.getDate();

	return [this.getFullYear(),
			(mm>9 ? '' : '0') + mm,
			(dd>9 ? '' : '0') + dd
		].join('');
};

/**
 * =============
 * = FUNCTIONS =
 **
 * Function to send a message to background.js.
 * @param {string} type
 * @param {object} object
 * @param {function} callback
 * @returns {*}
 **/
function sendBackgroundMessage(type, object, callback)
{
	chrome.runtime.sendMessage({type, object}, callback);
}

/**
 * Function to format a given date.
 * @source: https://stackoverflow.com/questions/23593052/format-javascript-date-as-yyyy-mm-dd
 * @param {string} argDate
 * @returns {string} retDate
 **/
function formatDate(argDate)
{
	let tDate = new Date(argDate),
		month = '' + (tDate.getMonth() + 1),
		day = '' + tDate.getDate(),
		year = tDate.getFullYear(),
		hours = '' + tDate.getHours(),
		minutes = '' + tDate.getMinutes();

	if (month.length < 2) month = '0' + month;
	if (day.length < 2) day = '0' + day;
	if (hours.length < 2) hours = '0' + hours;
	if (minutes.length < 2) minutes = '0' + minutes;

	return [month, day, year, hours, minutes].join('-');
}

/**
 * Function to check if this browser is Firefox.
 * @source: https://github.com/EFForg/https-everywhere/blob/bcaf7bdecf147c2a46d7ed73bce64389d828d865/chromium/background-scripts/background.js
 * @param {*}
 * @returns {boolean} 
 **/
function isFirefox()
{
	if (typeof(browser) != "undefined")
	{
		// #TODO: get information about browser properly.
		/*_getBrowserInfo().then(function(info) {
			if (info == "Firefox") {
				return true;
			} else {
				return false;
			}
		});*/
		return true;
	} else {
		return false;
	}
}

/**
 * =========================
 * = GET and SET functions =
 **
 **
 * Function to get current date and format it.
 * @param {*}
 * @returns {date}
 **/
function _getDate()
{
	let tDate = new Date();

	return tDate.yyyymmdd();
}

/**
 * Function to get stored value in BrowserStorage.
 * @param {string} argKey
 * @param {object} argStorageArea
 * @returns {promise} result
 **/
function _getStoredValues(argKey, argStorageArea)
{
	return new Promise(result => argStorageArea.get([argKey], data => result(data[argKey])));
}

/**
 * Async function to get user preferences.
 * @param {string} argKey
 * @returns {promise} result
 **/
async function _getUserSettings(argKey)
{
	return await _getStoredValues(argKey, chrome.storage.local);
}

/**
 * Function to get promise browser information.
 * @param {*}
 * @returns {promise} result
 **/
function _getBrowserGenericInfo()
{
	return new Promise( resolve => {
		let result = browser.runtime.getBrowserInfo();
		resolve(result);
	});
}

/**
 * Async function to get browser information.
 * @param {*}
 * @returns {object} result
 **/
async function _getBrowserInfo()
{
	return await _getBrowserGenericInfo();
}

tabs2text-settings.js
/************************
 ** TABS2TEXT
 **   description: Export all your opened tabs to a CSV, HTML, JSON or Markdown file.
 **   homepage_url: https://github.com/deniszanin/tabs2text
 **   license: Creative Commons Attribution-NonCommercial 4.0 International (CC BY-NC 4.0)
 **
 ** FILE: tabs2text-settings.js
 ************************
 **/
"use strict";

/**
 * ===================
 * = EVENT LISTENERS =
 *
 ** Default action on loading page.
 **/
document.addEventListener("DOMContentLoaded", () => {
	// Add an event listener "click" to SAVE PREFERENCES button.
	let h_btnSave = document.getElementById("tabs-saveSettings");
	h_btnSave.addEventListener("click", clickSavePreferences);

	// Get stored values when loading settings page:
	_getUserSettings("storCurrentWindow").then(function(resCurrentWindow) {
		if (typeof resCurrentWindow != "undefined") {
			if (resCurrentWindow) 
				document.getElementById("export_allTabs_no").checked = true;
			else
				document.getElementById("export_allTabs_yes").checked = true;
		} else {
			document.getElementById("export_allTabs_no").checked = true;
			sendBackgroundMessage("debug_log_on", "User settings (storCurrentWindow) is empty (setting default values).");
		}
	});

	_getUserSettings("storFileFormat").then(function(resFileFormat) {
		if (resFileFormat) {
			_setUIFileFormat(resFileFormat);
		} else {
			sendBackgroundMessage("debug_log_on", `User settings (storFileFormat) is empty (setting default values).`);
		}
	});
});

/**
 * =============
 * = FUNCTIONS =
 **
 * Function to save user preferences to BrowserStorage.
 * @param {*}
 * @return {*}
 **/
function clickSavePreferences()
{
	// Preference options to be saved.
	const userPreferences = {
		storCurrentWindow: _getUICurrentWindow(),
		storFileFormat: _getUIFileFormat()
	};

	chrome.storage.local.set(userPreferences, function() {
		document.getElementById("spn_save_status").innerText = chrome.i18n.getMessage("lbl_save_preferences_action");
		sendBackgroundMessage("debug_log_on", `User preferences: ${JSON.stringify(userPreferences)}.`);
	});
}

/**
 * =========================
 * = GET and SET functions =
 **
 * Function to get UI file format options.
 * @param {*}
 * @returns {string} retFormat
 **/
function _getUIFileFormat()
{
	let objOptions = document.getElementsByName("format_file_radio");
	let intOptions = objOptions.length;
	let retFormat;

	for (let c=0;c < intOptions; c++) {
		if (objOptions[c].checked == true) retFormat = objOptions[c].value;
	}

	return retFormat;
}

/** 
 * Function to get UI option for current window.
 * @param {*}
 * @returns {boolean} retCurrentWindow
 **/
function _getUICurrentWindow()
{
	let objALLTabs = document.getElementsByName("export_allTabs_radio");
	let intALLTabs = objALLTabs.length;
	let retCurrentWindow;

	for (let c=0;c < intALLTabs; c++) {
		if (objALLTabs[c].checked == true) {
			retCurrentWindow = (objALLTabs[c].value === "true");
		} 
	}

	return retCurrentWindow;
}

/**
 * Function to set UI file format options.
 * @param {string} argFileFormat
 * @returns {*}
 **/
function _setUIFileFormat(argFileFormat)
{
	let objOptions = document.getElementsByName("format_file_radio");
	let intOptions = objOptions.length;

	for (let c=0;c < intOptions; c++) {
		if (objOptions[c].value == argFileFormat) objOptions[c].checked = true;
	}
}

//EOF

tabs2text-popup.js
/************************
 ** TABS2TEXT
 **   description: Export all your opened tabs to a CSV, HTML, JSON or Markdown file.
 **   homepage_url: https://github.com/deniszanin/tabs2text
 **   license: Creative Commons Attribution-NonCommercial 4.0 International (CC BY-NC 4.0)
 **
 ** FILE: tabs2text-popup.js
 ************************
 **/
"use strict";

/**
 * =====================
 * = GLOBALS VARIABLES =
 **/
const DEFAULT_FILE_PREFIX = "tabs2text";
const DEFAULT_GITHUB_PAGE = "https://github.com/deniszanin/tabs2text";

// Default values
let g_isFirefox = false;
let g_fileExtension = "html";
let g_fileMimetype = "text/html;charset=utf=8";
let g_objJSONContent = [];

/**
 * ===================
 * = EVENT LISTENERS =
 *
 ** Default action on loading page.
 **/
document.addEventListener("DOMContentLoaded", () => {
	// Add an event listener "click" to GENERATE button.
	let h_btnGenerate = document.getElementById("tabs-generateFile");
	h_btnGenerate.addEventListener("click", function() {
		exportFileContent(_getUIFileFormat(), _getUICurrentWindow());
	}, false);

	// Add an event listener "click" to PREFERENCES button.
	let h_btnPreferences = document.getElementById("tabs-openPreferences");
	h_btnPreferences.addEventListener("click", function() {
		let tabPreferences = chrome.tabs.create({
			url: chrome.extension.getURL("tabs2text-settings.html")
		});
	}, false);

	//////
	////// LOADING SETTINGS

	// Update user interface version and icon.
	_setUIVersion();
	_setUIIcon();

	// Get which browser is running.
	g_isFirefox = isFirefox();

	// Get stored values when loading settings page:
	_getUserSettings("storCurrentWindow").then(function(resCurrentWindow) {
		if (typeof resCurrentWindow != "undefined") {
			if (resCurrentWindow) 
				document.getElementById("export_allTabs_no").checked = true;
			else
				document.getElementById("export_allTabs_yes").checked = true;
		} else {
			document.getElementById("export_allTabs_no").checked = true;
			sendBackgroundMessage("debug_log_on", "User settings (storCurrentWindow) is empty (setting default values).");
		}
	});

	_getUserSettings("storFileFormat").then(function(resFileFormat) {
		if (resFileFormat) {
			_setUIFileFormat(resFileFormat);
		} else {
			sendBackgroundMessage("debug_log_on", `User settings (storFileFormat) is empty (setting default values).`);
		}
	});
});

/**
 * =============
 * = FUNCTIONS =
 **
 * Function to export tab information to a file.
 * @param {string} argFileFormat
 * @param {boolean} argCurrentWindow
 * @returns {*}
 **/
function exportFileContent(argFileFormat, argCurrentWindow)
{
	let queryOptions;

	if (argFileFormat == "") throw new Error ("Error: invalid exportFile parameter.");

	_setFileMIMEExtension(argFileFormat);

	if (argCurrentWindow) {
		queryOptions = {
			currentWindow: true
		};
	} else {
		queryOptions = {};
	}

	chrome.tabs.query(queryOptions, objALLTabs => {
		let idWindow = "allwindows";
		let tmpContent = "";
		let tmpCounter = 1;

		for (let currentTab of objALLTabs) {
			tmpContent = tmpContent + formatContent(argFileFormat, tmpCounter, currentTab.title, currentTab.url, formatDate(currentTab.lastAccessed));
			tmpCounter += 1;

			if (argCurrentWindow) idWindow = currentTab.windowId;
		}

		let fileContent = formatHeader(argFileFormat);
		fileContent = fileContent + tmpContent;

		if (argFileFormat == "JSON") {
			fileContent = JSON.stringify(g_objJSONContent, null, 4);
		} else if (argFileFormat == "HTML") {
			fileContent = fileContent + "</DL><p>";
		}

		// Create a BLOB file with MIME type (Chrome doesn't understand without it).
		let fileBlob = new Blob([fileContent], { type: g_fileMimetype });
		let tmpURL = (window.webkitURL || window.URL).createObjectURL(fileBlob);

		let nowDate = _getDate();
		let expFilename = DEFAULT_FILE_PREFIX + "-" + nowDate + "-" + idWindow + "." + g_fileExtension;

		// Send temporary file blob URL to "background.js".
		sendFileContentToDownloader(tmpURL, expFilename, fileBlob);
	});
}

/**
 * Function to format the header of the file.
 * @param {string} argFormat
 * @returns {string} returnContent
 **/
function formatHeader(argFormat)
{
	let returnContent = "";

	if (argFormat == "CSV") {
		returnContent = "id;title;url";
		if (g_isFirefox) returnContent = returnContent + ";last access";
		returnContent = returnContent + "\n";
	} else if (argFormat == "HTML") {
		returnContent = "<!DOCTYPE NETSCAPE-Bookmark-file-1>\n";
		returnContent = returnContent + "<!-- This is an automatically generated file.\n";
		returnContent = returnContent + "Generated by Tabs2Text (" + DEFAULT_GITHUB_PAGE + ") ! -->\n";
		returnContent = returnContent + "<META HTTP-EQUIV=\"Content-Type\" CONTENT=\"text/html; charset=UTF-8\">\n";
		returnContent = returnContent + "<TITLE>Tabs2Text links</TITLE>\n";
		returnContent = returnContent + "<H1>Tabs2Text links</H1>\n";
		returnContent = returnContent + "<DL><p>\n";
	} else if (argFormat == "MD") {
		returnContent = "TABS2TEXT LINKS\n";
		returnContent = returnContent + "================\n";
		returnContent = returnContent + "# LINKS\n\n";
		returnContent = returnContent + "## Generated by [Tabs2Text](" + DEFAULT_GITHUB_PAGE + ")\n\n";
	}

	return returnContent;
}

/**
 * Function to format the content of tabs, which will be exported to the file.
 * @param {string} argFormat
 * @param {int} argID
 * @param {string} argTitle
 * @param {string} argURL
 * @param {string} argLastAccessed
 **/
function formatContent(argFormat, argID, argTitle, argURL, argLastAccessed)
{
	let returnContent = "";
	let t_objContent = {};

	if (argFormat == "CSV") {
		//returnContent = argID + ";" + argTitle + ";" + argURL + ";" + argLastAccessed + "\n";
		returnContent = argID + ";";
		returnContent = returnContent + argTitle + ";";
		returnContent = returnContent + argURL;
		if (g_isFirefox) returnContent = returnContent + ";" + argLastAccessed;
		returnContent = returnContent + "\n";
	} else if (argFormat == "HTML") {
		returnContent = "\t<DT><A HREF=\"" + argURL + "\">" + argID + ". " + argTitle;
		if (g_isFirefox) returnContent = returnContent + " (last access: " + argLastAccessed + ")";
		returnContent = returnContent + "</A>\n";
	} else if (argFormat == "MD") {
		returnContent = "### " + argID + ". [" + argTitle + "](" + argURL + ")";
		if (g_isFirefox) returnContent = returnContent + ", last access: " + argLastAccessed;
		returnContent = returnContent + "\n";
	} else if (argFormat == "JSON") {
		t_objContent.id = argID;
		t_objContent.title = argTitle;
		t_objContent.url = argURL;
		if (g_isFirefox) t_objContent.last_access = argLastAccessed;

		g_objJSONContent.push(t_objContent);
	}

	return returnContent;
}

/**
 * ===================
 * = EXTERNAL NOTIFY =
 **
 **
 * Function to send file content information to "background.js".
 * @param {string} argURL
 * @param {string} argFilename
 * @param {blob} argBlob
 * @returns {*}
 */
function sendFileContentToDownloader(argURL, argFilename, argBlob)
{
	sendBackgroundMessage("to_downloader", {"url": argURL, "filename": argFilename, "blob": argBlob});
}

/**
 * =========================
 * = GET and SET functions =
 **
 * Function to get UI file format options.
 * @param {*}
 * @returns {string} retFormat
 **/
function _getUIFileFormat()
{
	let objOptions = document.getElementsByName("format_file_radio");
	let intOptions = objOptions.length;
	let retFormat;

	for (let c=0;c < intOptions; c++) {
		if (objOptions[c].checked == true) retFormat = objOptions[c].value;
	}

	return retFormat;
}

/** 
 * Function to get UI option for current window.
 * @param {*}
 * @returns {boolean} retCurrentWindow
 **/
function _getUICurrentWindow()
{
	let objALLTabs = document.getElementsByName("export_allTabs_radio");
	let intALLTabs = objALLTabs.length;
	let retCurrentWindow;

	for (let c=0;c < intALLTabs; c++) {
		if (objALLTabs[c].checked == true) {
			retCurrentWindow = (objALLTabs[c].value === "true");
		} 
	}

	return retCurrentWindow;
}

/**
 * Function to set UI file format options.
 * @param {string} argFileFormat
 * @returns {*}
 **/
function _setUIFileFormat(argFileFormat)
{
	let objOptions = document.getElementsByName("format_file_radio");
	let intOptions = objOptions.length;

	for (let c=0;c < intOptions; c++) {
		if (objOptions[c].value == argFileFormat) objOptions[c].checked = true;
	}
}

/**
 * Function to set the current version of this extension (addon) in UI.
 * @param {*}
 * @returns {*}
 **/
function _setUIVersion()
{
	let locManifest = chrome.runtime.getManifest();
	let spnVersion = document.getElementById("v-current-version");
	spnVersion.innerText = locManifest.version;
}

/**
 * Function to set the icon of this extension (addon) in UI.
 * @param {*}
 * @returns {*}
 **/
function _setUIIcon()
{
	let locManifest = chrome.runtime.getManifest();
	let srcIcon = document.getElementById("ico-tabs2text");
	srcIcon.src = locManifest.icons["48"];
}

/**
 * Function to set file extension and MIME type.
 * @source https://www.iana.org/assignments/media-types/media-types.xhtml
 * @param {string} argFormat
 * @returns {*}
 **/
function _setFileMIMEExtension(argFormat)
{
	if (argFormat == "CSV") {
		g_fileExtension = "csv";
		g_fileMimetype = "text/csv;charset=utf=8";
	} else if (argFormat == "HTML") {
		g_fileExtension = "html";
		g_fileMimetype = "text/html;charset=utf=8";
	} else if (argFormat == "MD") {
		g_fileExtension = "md";
		g_fileMimetype = "text/markdown;charset=utf=8";
	} else if (argFormat == "JSON") {
		g_fileExtension = "json";
		g_fileMimetype = "application/json;charset=utf=8";
	} else {
		g_fileExtension = "txt";
		g_fileMimetype = "text/plain;charset=utf=8";
	}
}

 //EOF